Add CoSERV API implementation

thomas-fossati opened the pull request:
This PR introduces `coserv`, a new service endpoint that implements the CoSERV API described in [draft-ietf-rats-coserv-02](https://www.ietf.org/archive/id/draft-ietf-rats-coserv-02.html).

setrofim commented on coserv/api/handler.go:
```diff
@@ -0,0 +1,169 @@
+	EndorsementDistibutor endorsementdistributor.IEndorsementDistributor
+}
+
+func NewHandler(endorsementdistributor endorsementdistributor.IEndorsementDistributor, logger *zap.SugaredLogger) Handler {
```
```suggestion
func NewHandler(
	endorsementdistributor endorsementdistributor.IEndorsementDistributor,
	logger *zap.SugaredLogger,
) Handler {
```

thomas-fossati replied:
Addressed in [7bcdc10](https://github.com/veraison/services/pull/360/commits/7bcdc105cc1bd48a395777dc87d10bcfef476e18)

setrofim commented on vts/coservsigner/icoservsigner.go:
```diff
@@ -0,0 +1,17 @@
+)
+
+type ICoservSigner interface {
+	Init(cfg Cfg, fs afero.Fs) error
```
Should the `Init` be part of the interface? Surely, this will be specific to the implementation? (E.g. some other implementation may not rely on a file system, and won't take an `afero.Fs`).

thomas-fossati replied:
The reason why `afero.Fs` pops up here is because I am reusing the earsigner’s `KeyLoader`, which takes it to instantiate a new `KeyLoader` from which to load by URL.

thomas-fossati commented on coserv/api/handler.go:
```diff
@@ -0,0 +1,169 @@
+
+func (o Handler) CoservRequest(c *gin.Context) {
+	offered := c.NegotiateFormat(CoservMTs...)
+	if offered != CoservMTs[0] && offered != CoservMTs[1] {
```
The number of media types is not going to increase in the future.  There’s only going to be a signed and unsigned version of the coserv type.
Your suggestion is more elegant, though.

thomas-fossati replied:
Addressed in [7bcdc10](https://github.com/veraison/services/pull/360/commits/7bcdc105cc1bd48a395777dc87d10bcfef476e18)

thomas-fossati commented on vts/trustedservices/trustedservices_grpc.go:
```diff
@@ -553,6 +620,156 @@ func (o *GRPC) GetEARSigningPublicKey(context.Context, *emptypb.Empty) (*proto.P
+func (o *GRPC) GetCoservSigningPublicKey(context.Context, *emptypb.Empty) (*proto.PublicKey, error) {
+	// If CoSERV is not enabled, return an error.
+	if o.CoservSigner == nil {
+		return &proto.PublicKey{Key: ""}, nil
```
Thanks for catching this.  It’s a leftover from a previous iteration of the interface.

thomas-fossati commented on scheme/common/scripts/gen-corim:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/opt/homebrew/bin/bash
```
Addressed in [7bcdc10](https://github.com/veraison/services/pull/360/commits/7bcdc105cc1bd48a395777dc87d10bcfef476e18)

thomas-fossati commented on coserv/cmd/coserv-service/main.go:
```diff
@@ -0,0 +1,113 @@
+		log.Fatalf("Could not initialize VTS client: %v", err)
+	}
+
+	vtsState, err := vtsClient.GetServiceState(context.TODO(), &emptypb.Empty{})
```
Addressed in [43fc43d](https://github.com/veraison/services/pull/360/commits/43fc43d0163cf88f6e0c18d37c8e513b5a63e6af)

yogeshbdeshpande commented on coserv/cmd/coserv-service/Makefile:
```diff
@@ -0,0 +1,16 @@
+# Copyright 2023 Contributors to the Veraison project.
```
```suggestion
# Copyright 2025 Contributors to the Veraison project.
```

yogeshbdeshpande commented:
@thomas-fossati : Where is the README.md that defines:

This is a CoSERV Implementation as per the Draft: XYZ
This is applicable to fetch Endorsements in a specific format as defined in the draft.

yogeshbdeshpande commented on proto/endorsement_query.proto:
```diff
@@ -0,0 +1,22 @@
+
+message EndorsementQueryOut {
+  Status status = 1;
+  bytes result_set = 2;
```
would be good to specify like line 15,

is the result set just a CBOR byte buffer 
OR
B64 encoded?

yogeshbdeshpande commented on scheme/amd-kds-coserv/coserv_handler.go:
```diff
@@ -0,0 +1,150 @@
+)
+
+func constructVcekUrl(instance *coserv.StatefulInstance) string {
+	// TODO(paulhowardarm) - deduce the product name and TCB parameters from the
```
It would be really nice to have a github issue for Line 25 and Line 71 to track missing bits of requirement so that future work can be scoped out!

yogeshbdeshpande commented on scheme/nvidia-coserv/coserv_handler.go:
```diff
@@ -0,0 +1,225 @@
+
+func (s CoservProxyHandler) addReferenceValuesForClass(query *coserv.Query, c *coserv.StatefulClass, results *coserv.ResultSet) error {
+	if *c.Class.Vendor != "NVIDIA" {
+		return fmt.Errorf("vendors other than NVIDIA not supported by this proxy plug-in")
```
```suggestion
		return errors.New("vendors other than NVIDIA not supported by this proxy plug-in")
```

yogeshbdeshpande commented on scheme/nvidia-coserv/coserv_handler.go:
```diff
@@ -0,0 +1,225 @@
+	}
+
+	if c.Class.Model == nil {
+		return fmt.Errorf("no Model field supplied in NVIDIA class environment")
```
```suggestion
		return errors.New("no Model field supplied in NVIDIA class environment")
```

thomas-fossati commented on scheme/nvidia-coserv/coserv_handler.go:
```diff
@@ -0,0 +1,225 @@
+	rimServiceResponse, err := callRimService(rimid)
+
+	if err != nil {
+		return err
```
I think it’s fine to return whatever NV API gives us.

yogeshbdeshpande commented on scheme/nvidia-coserv/coserv_handler.go:
```diff
@@ -0,0 +1,225 @@
+			return fmt.Errorf("CoSERV proxy plug-in cannot produce collected artifacts for non-CORIM format %s", rimServiceResponse.RimFormat)
+		}
+
+		// We now know that the format is CORIM, so it should be possible to decode as such
```
An architectural point, worth discussing via a github issue, may be I will track it in the repo..

If NVIDIA is already reporting a CoRIM which is signed, I see no point breaking it down, re-formatting it.

Also potentially I would be worried, even more, if we lose some information from CoRIM in the response!

Yes, there is a facility to stash the entire blob as a side object to the re-package, if I understand CoSERV correctly, then however, this sounds like an over-kill !

thomas-fossati replied:
There is no architectural issue here; this is exactly how CoSERV is supposed to work.  It's more of a RATS discussion around the CoSERV spec that should be done over the RATS ML.

yogeshbdeshpande commented on scheme/amd-kds-coserv/scheme.go:
```diff
@@ -0,0 +1,12 @@
+
+var (
+	CoservMediaTypes = []string{
+		`application/coserv+cbor; profile="tag:github.com/veraison,2023:amd_kds_coserv_proxy#1.0.0"`,
```
No but that reflects the feature functionality year so makes sense :-)

yogeshbdeshpande commented on vtsclient/vtsclient_grpc.go:
```diff
@@ -195,6 +195,54 @@ func (o *GRPC) GetEARSigningPublicKey(ctx context.Context, in *emptypb.Empty, op
 	return c.GetEARSigningPublicKey(ctx, in, opts...)
 }
 
+func (o *GRPC) GetCoservSigningPublicKey(ctx context.Context, in *emptypb.Empty, opts ...grpc.CallOption) (*proto.PublicKey, error) {
```
the name of the API should be
GetCoservVerificationPublicKey OR GetCoservVerificationKey

thomas-fossati replied:
Why? It’s the public key associated with the CoSERV signing key … Note that the naming scheme is consistent with that used for the EAR signing key.

yogeshbdeshpande commented on vts/trustedservices/trustedservices_grpc.go:
```diff
@@ -553,6 +620,158 @@ func (o *GRPC) GetEARSigningPublicKey(context.Context, *emptypb.Empty) (*proto.P
+	if err == nil {
+		// No error means we have a proxy plugin, so delegate to that.
+		out, err = o.getEndorsementsFromProxy(handlerPlugin, query)
+	} else {
```
A more correct implementation is to check for a specific error that leads to conclusion that there is no suitable proxy plugin

Then, only cater to this error, other wise blind assumption can lead to problems down the line!

yogeshbdeshpande commented:
> > @thomas-fossati I see a lot of new .cbor files in the PR, under scheme/psa-iot/test/corim and similarly in other schemes, like psa-iot
> > However I do not see any src/ folder for this, am I missing something, OR there is no change?
> 
> There is no change in the CoRIM JSON templates. Only their serialisations (i.e., the CBOR files) have been updated to match CoRIM.

ok, so I assume, it is due to new revision of CoRIM that is been applied. ..?
However, why did the CBOR changed, unless anything is changed, I am confused.

yogeshbdeshpande commented:
> > Approval subject to all Pending comments been resolved!
> 
> Thanks for the review! I have addressed all your comments and raised a new issue in the veraison/docs repo to deal with the documentation.

Yes, it is important that the main page in Veraison README also reflects the changes, which is also lacking at the moment

thomas-fossati commented:
> > > Approval subject to all Pending comments been resolved!
> > 
> > 
> > Thanks for the review! I have addressed all your comments and raised a new issue in the veraison/docs repo to deal with the documentation.
> 
> Yes, it is important that the main page in Veraison README also reflects the changes, which is also lacking at the moment

Feel free to extend https://github.com/veraison/docs/issues/71 with any other related documentation task.
